### DISCUSSION TITLE
where and how is the node encryption (by password) feature implemented

### DISCUSSION THREAD

**Discussion by nnako** [freeplane/freeplane#2127] [Help]
Within the Freeplane editor, users can encrypt and decrypt nodes (including possible further child noodes). Within the mindmap's `XML` file these encrypted nodes are represented by a BASE64 value, like:

```xml
<node TEXT="pers" STYLE_REF="klein und grau" ENCRYPTED_CONTENT="fUg00upinp0=m9wdrklgLvVa3SKgget9xzW+6OPALUJ4bK18FeBcGUG9rg1qC0wqKbjJdUhjHqXwYnn6xcJBN5++Z0AL3i6sPvEjvg5L1cF6qZKv2sX/p1Q33xA6lb1G3gvGREg7fVSzwrLs8T0yiMUqFv0/K9Vj0fgoa4ArKYmVa4nd9Jh34ntNiAlDj8GUmOnLwQRs/8I00ax/DByMAjhJ5dGVYKR4i6Ni3zabaqyarKvjKbO0vDQj1kGEczJ9O2Jj/tl2xM3D0J8bCfaRJUyaTcgn+uI81RRXpFz+pmVo0gZ813KdCWeeJLd8fqrHPZeWUjrcAgG/Ul8UapwQ7KPzFUQQrDefTR0hWJr9YGahqYibTllDveT63iezdHxQWoLUmbmktLtY00gSKqQli1kQSsRXM0cw4Kcn5KnfqKUXfAcJUc61PgidnGAAqA1o9FR/kO5Y0JKR" ID="ID_1472189397" CREATED="1732537044960" MODIFIED="1732537046620"/>
```

I guess, the BASE64 encoded portion visible as a value of the attribute `ENCRYPTED_CONTENT` is the result of some standard encryption process mentioned by @macmarrum in https://github.com/freeplane/freeplane/discussions/1807#discussioncomment-9271272 .

Can someone point me to some information about **where** and possibly **how** this encryption and decryption feature is implemented within the Freeplane code base? Further details about a necessary processing sequence, used encryption concepts like `MD5` and `DES`, padding configurations, ... would be helpful.

Maybe I could work on (external) ways to provide help when somebody has lost a password, as mentioned in https://github.com/freeplane/freeplane/discussions/2123#discussion-7538051 by @ebhdoc .

Thanks.

**Comment by macmarrum** [comment]
Here's how I go about finding the code responsible for an action in Freeplane:

1. Launch `Tools->Developer Tools->Menu item info` (part of the add-on [Developer Tools](https://docs.freeplane.org/getting-started/Add-ons_(install).html#for-developers))
2. Select the action you are interested in (e.g. `Tools->Enter password`)
3. Copy/note the value of `Key:`
4. Search for that value in Freeplane's source code -> you'll find a class that extends `AFreeplaneAction`
5. Review the `actionPerformed` method and any code that it references
